sync/src/relayer/compact_block_process.rs

```diff
@@ -273,3 +156,182 @@ impl<'a> HeaderProvider for CompactBlockMedianTimeView<'a> {
         (self.fn_get_pending_header)(hash.to_owned())
     }
 }
+
+// check compact block's uncles and proposals length
+// check compact block height
+fn non_contextual_check(
+    compact_block: &CompactBlock,
+    consensus: &Consensus,
+    active_chain: &ActiveChain,
+) -> Status {
+    if compact_block.uncles().len() > consensus.max_uncles_num() {
+        return StatusCode::ProtocolMessageIsMalformed.with_context(format!(
+            "CompactBlock uncles count({}) > consensus max_uncles_num({})",
+            compact_block.uncles().len(),
+            consensus.max_uncles_num()
+        ));
+    }
+    if (compact_block.proposals().len() as u64) > consensus.max_block_proposals_limit() {
+        return StatusCode::ProtocolMessageIsMalformed.with_context(format!(
+            "CompactBlock proposals count({}) > consensus max_block_proposals_limit({})",
+            compact_block.proposals().len(),
+            consensus.max_block_proposals_limit(),
+        ));
+    }
+
+    // Only accept blocks with a height greater than tip - N
+    // where N is the current epoch length
+    let header = compact_block.header().into_view();
+    let block_hash = header.hash();
+    let tip = active_chain.tip_header();
+    let epoch_length = active_chain.epoch_ext().length();
+    let lowest_number = tip.number().saturating_sub(epoch_length);
+
+    if lowest_number > header.number() {
+        return StatusCode::CompactBlockIsStaled.with_context(block_hash);
+    }
+
+    Status::ok()
+}
+
+// check compact block if already stored in db
+// check compact block extension validation
+// check compact block's parent block is not stored in db
+// check compact block is in pending
+// check compact header verification
+fn contextual_check(
+    compact_block: &CompactBlock,
+    shared: &Arc<SyncShared>,
+    active_chain: &ActiveChain,
+    nc: &Arc<dyn CKBProtocolContext>,
+    peer: PeerIndex,
+) -> Status {
+    let header = compact_block.header().into_view();
```

Comment:
Don't always `into_view`, it's expensive(always hash itself), header must `into_view` in global, and always use reference on it

Reply:
Thank, i fixed it as your suggestion.